stdlib/public/core/CollectionDifference.swift

```diff
@@ -0,0 +1,353 @@
+  }
+  
+  public func index(before index: CollectionDifference<ChangeElement>.Index) -> CollectionDifference<ChangeElement>.Index {
+    return Index(index.i - 1)
+  }
+  
+  public func formIndex(_ index: inout CollectionDifference<ChangeElement>.Index, offsetBy distance: Int) {
+    index = Index(index.i + distance)
+  }
+  
+  public func distance(from start: CollectionDifference<ChangeElement>.Index, to end: CollectionDifference<ChangeElement>.Index) -> Int {
+    return end.i - start.i
+  }
+}
+
+extension CollectionDifference.Change: Equatable where ChangeElement: Equatable {}
+
+extension CollectionDifference: Equatable where ChangeElement: Equatable {}
+
+extension CollectionDifference.Change: Hashable where ChangeElement: Hashable {}
+
+extension CollectionDifference: Hashable where ChangeElement: Hashable {
+
+  /// Infers which `ChangeElement`s have been both inserted and removed only
+  /// once and returns a new difference with those associations.
+  ///
+  /// - Returns: an instance with all possible moves inferred.
+  ///
+  /// - Complexity: O(*n*) where *n* is `self.count`
+  public func inferringMoves() -> CollectionDifference<ChangeElement> {
+    let removeDict: [ChangeElement:Int?] = {
+      var res = [ChangeElement:Int?](minimumCapacity: Swift.min(removals.count, insertions.count))
+      for r in removals {
+        let element = r.element
+        if res[element] != .none {
+          res[element] = .some(.none)
+        } else {
+          res[element] = .some(r.offset)
+        }
+      }
+      return res.filter { (_, v) -> Bool in v != .none }
+    }()
+    
+    let insertDict: [ChangeElement:Int?] = {
+      var res = [ChangeElement:Int?](minimumCapacity: Swift.min(removals.count, insertions.count))
+      for i in insertions {
+        let element = i.element
+        if res[element] != .none {
+          res[element] = .some(.none)
+        } else {
+          res[element] = .some(i.offset)
+        }
+      }
+      return res.filter { (_, v) -> Bool in v != .none }
+    }()
+
+    return CollectionDifference.init(validatedChanges:map({ (c: CollectionDifference<ChangeElement>.Change) -> CollectionDifference<ChangeElement>.Change in
+      switch c {
+      case .remove(offset: let o, element: let e, associatedWith: _):
+        if removeDict[e] == nil {
+          return c
+        }
+        if let assoc = insertDict[e] {
+          return .remove(offset: o, element: e, associatedWith: assoc)
+        }
+      case .insert(offset: let o, element: let e, associatedWith: _):
+        if insertDict[e] == nil {
+          return c
+        }
+        if let assoc = removeDict[e] {
+          return .insert(offset: o, element: e, associatedWith: assoc)
+        }
+      }
+      return c
+    }))
+  }
+}
+
+extension CollectionDifference.Change: Codable where ChangeElement: Codable {
+  private enum CodingKeys: String, CodingKey {
```

Comment:
```suggestion
  private enum _CodingKeys: String, CodingKey {
```

(Foundation isn't pedantic about using the underscores, but the stdlib core is, and AFAIK the type name is not required to literally be `CodingKeys`.)